File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/kafka/producer/SendNotificationProducer.java
package com.sleepkqq.sololeveling.player.service.kafka.producer;

import static com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics.SEND_NOTIFICATION_TOPIC;

import com.sleepkqq.sololeveling.avro.notification.SendNotificationEvent;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class SendNotificationProducer {

  private final KafkaTemplate<String, SendNotificationEvent> kafkaTemplate;

  public void send(SendNotificationEvent event) {
    kafkaTemplate.send(SEND_NOTIFICATION_TOPIC, event);
    log.info(">> Notification sent | transactionId={}", event.getTransactionId());
  }
}

File: solo-leveling-player-service/src/main/kotlin/com/sleepkqq/sololeveling/player/service/kafka/producer/SendNotificationProducer.kt
package com.sleepkqq.sololeveling.player.service.kafka.producer

import com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics
import com.sleepkqq.sololeveling.avro.notification.SendNotificationEvent
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.retry.annotation.Backoff
import org.springframework.retry.annotation.Retryable
import org.springframework.stereotype.Service

@Service
class SendNotificationProducer(
	private val kafkaTemplate: KafkaTemplate<String, SendNotificationEvent>
) {

	private val log = LoggerFactory.getLogger(javaClass)

	@Retryable(maxAttempts = 3, backoff = Backoff(delay = 1000, multiplier = 2.0))
	fun send(event: SendNotificationEvent) {
		val txId = event.transactionId
		log.info("Sending notification event | txId={}", txId)

		try {
			kafkaTemplate.send(KafkaTaskTopics.SEND_NOTIFICATION_TOPIC, txId, event)
			log.info("<< Notification event sent successfully | txId={}", txId)

		} catch (e: Exception) {
			log.error("Failed to send notification event | txId={}", txId, e)
			throw e
		}
	}
}
